### PR TITLE
kubeadm: default etcd address based on the IP family

### PR DESCRIPTION
**Pull request by aojea**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

kubeadm always configured the etcd endpoint to the IPv4 localhost address.

This patchs configures the right localhost address based on the used
AdvertiseAddress IP family.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: uses the apiserver AdvertiseAddress IP family to choose the etcd endpoint IP family for non external etcd clusters
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by aojea** [comment]
/hold
want to verify first is working correctly with ipv6
/test pull-kubernetes-e2e-kind-ipv6
/assign @neolit123 @fabriziopandini @ereslibre

**Review by neolit123 — COMMENTED** [review]
thanks @aojea 
/approve
/assign @ereslibre

**Comment by aojea** [comment]
@BenTheElder is it possible that the nodes running the IPv6 jobs are influencing the tests?
Before this PR the apiserver and etcd were communicating using ipv4 with a reasonable flakiness and a small number of jobs filing. Enabling IPv6 between the apiserver and etcd it fails constantly with more than 300 tests failing

File: cmd/kubeadm/app/phases/etcd/local.go [hunk 1]
```diff
@@ -176,7 +179,11 @@ func GetEtcdPodSpec(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.A
 		etcdVolumeName:  staticpodutil.NewVolume(etcdVolumeName, cfg.Etcd.Local.DataDir, &pathType),
 		certsVolumeName: staticpodutil.NewVolume(certsVolumeName, cfg.CertificatesDir+"/etcd", &pathType),
 	}
+	// TODO: probeHostname is obtained from listen-metrics-urls but the argument is created later with getEtcdCommand
```
Comment (aojea) [inline cmd/kubeadm/app/phases/etcd/local.go]: I was thinking in different ways of dealing with this chicken an egg problem but didn't find a better one. Will nice some guidance here, just choose the simplest solution

**Comment by aojea** [comment]
/test pull-kubernetes-conformance-kind-ipv6
flaky, only failed 3 tests, seems that now ipv6 is healthier

**Comment by aojea** [comment]
/test pull-kubernetes-e2e-gce
doesn't seems related to this PR

**Comment by aojea** [comment]
/hold cancel
this is ready for review
It improves the e2e kind ipv6 tests, right now the kube-proxy and other daemon are less verbose with this change, on theory is that maybe having apiserver - etcd communicating in ipv4 and apiserver-kubelet in ipv6 was adding a penalty or a problem
/cc @BenTheElder

**Comment by aojea** [comment]
/test pull-kubernetes-conformance-kind-ipv6
reported flaky test https://github.com/kubernetes/kubernetes/issues/85762

File: cmd/kubeadm/app/phases/etcd/local.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/etcd/local.go]: GetEtcdProbeEndpoint() can be made to have a bool flag for ipv6:

```
probeHostname, probePort, probeScheme := staticpodutil.GetEtcdProbeEndpoint(&cfg.Etcd, utilsnet.IsIPv6String(endpoint.AdvertiseAddress))

GetEtcdProbeEndpoint(cfg *kubeadmapi.Etcd, bool isIPv6) (string, int, v1.URIScheme)
```

this is better than the override after the function call, but requires minor unit test changes.

Comment (aojea) [reply cmd/kubeadm/app/phases/etcd/local.go]: done

File: cmd/kubeadm/app/util/staticpod/utils_test.go [hunk 2]
```diff
@@ -131,6 +150,7 @@ func TestGetEtcdProbeEndpoint(t *testing.T) {
 						"listen-metrics-urls": "https://1.2.3.4:1234,https://4.3.2.1:2381"},
 				},
 			},
+			isIPv6:           false,
```
Comment (neolit123) [inline cmd/kubeadm/app/util/staticpod/utils_test.go]: nit: `false` is the default value for a boolean struct member.
but this is fine like that too.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 3]
```diff
@@ -162,7 +162,12 @@ func getAPIServerCommand(cfg *kubeadmapi.ClusterConfiguration, localAPIEndpoint
 		}
 	} else {
 		// Default to etcd static pod on localhost
-		defaultArguments["etcd-servers"] = fmt.Sprintf("https://127.0.0.1:%d", kubeadmconstants.EtcdListenClientPort)
+		// localhost IP family should be the same that the AdvertiseAddress
+		etcdLocalhostAddress := "127.0.0.1"
+		if utilsnet.IsIPv6String(localAPIEndpoint.AdvertiseAddress) {
+			etcdLocalhostAddress = "::1"
```
Comment (ereslibre) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Our etcd certificate allows `localhost`, wouldn't it be better to use it instead?

Comment (aojea) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: localhost is a lottery, you don't known in advance what ip is going to resolve, check the following gist explaining the topic
https://gist.github.com/aojea/94f6f483173641647c731f582e52f0b0

Comment (ereslibre) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Yes, it's better to use IP addresses directly. I was thinking that using `localhost` we would rely on whatever has been set on the containers `/etc/hosts`, but that's just shifting the problem somewhere else.

**Review by ereslibre — APPROVED** [review]
Thank you @aojea.

/lgtm
/hold
^ for more eyes to lgtm.

**Comment by aojea** [comment]
The ipv6 job is flaky and only two storage tests failed, is ok, thanks for the quick reviews

**Pull request merged**